src/pharmpy/modeling/eta_additions.py

```diff
@@ -156,44 +166,163 @@ def add_iov(model, occ, list_of_parameters=None, eta_names=None):
+
+        if not all(
+            map(
+                lambda x: isinstance(x, list) and all(map(lambda y: isinstance(y, str), x)),
+                list_of_parameters,
+            )
+        ):
+            raise ValueError('not all parameters are strings')
+
+        etas = [_get_etas(model, grp, include_symbols=True) for grp in list_of_parameters]
+
+        for dist, grp in zip(etas, list_of_parameters):
+            assert len(dist) <= len(grp)
+            if len(dist) != len(grp):
+                raise ValueError(f'Some ETA in {dist} was given twice.')
+
     categories = _get_occ_levels(model.dataset, occ)
 
-    if eta_names and len(eta_names) != len(etas) * len(categories):
+    if eta_names and len(eta_names) != sum(map(len, etas)) * len(categories):
         raise ValueError(
-            f'Number of provided names incorrect, need {len(etas) * len(categories)} names.'
+            'Number of given eta names is incorrect, '
+            f'need {sum(map(len,etas)) * len(categories)} names.'
         )
-    elif len(categories) == 1:
+
+    if len(categories) == 1:
         raise ValueError(f'Only one value in {occ} column.')
 
-    iovs, etais = ModelStatements(), ModelStatements()
-    for i, eta in enumerate(etas, 1):
-        omega_name = str(next(iter(eta.sympy_rv.pspace.distribution.free_symbols)))
-        omega = S(f'OMEGA_IOV_{i}')  # TODO: better name
-        pset.append(Parameter(str(omega), init=pset[omega_name].init * 0.1))
+    # NOTE This declares the ETAS and their corresponding OMEGAs
+    if distribution == 'same-as-iiv':
+        # NOTE We filter existing IIV distributions for selected ETAs and then
+        # let the mixed distribution logic handle the rest
+        assert len(etas) == 1
+        etas_set = set(etas[0])
+        etas = []
+        for variables, dist in model.random_variables.distributions():
 
-        iov = S(f'IOV_{i}')
+            intersection = list(filter(etas_set.__contains__, variables))
 
-        values, conditions = [], []
+            if not intersection:
+                continue
 
-        for j, cat in enumerate(categories, 1):
-            if eta_names:
-                eta_name = eta_names[j - 1]
+            if len(variables) == 1:
+                assert isinstance(dist, NormalDistribution)
             else:
-                eta_name = f'ETA_IOV_{i}{j}'
+                assert isinstance(dist, MultivariateNormalDistribution)
+
+            etas.append(intersection)
 
-            eta_new = RandomVariable.normal(eta_name, 'iov', 0, omega)
-            rvs.append(eta_new)
+    # TODO: better names
+    def eta_name(i, k):
+        return eta_names[(i - 1) * len(categories) + k - 1] if eta_names else f'ETA_IOV_{i}{k}'
 
-            values += [S(eta_new.name)]
-            conditions += [Eq(cat, S(occ))]
+    def omega_iov_name(i, j):
+        return f'OMEGA_IOV_{i}' if i == j else f'OMEGA_IOV_{i}_{j}'
+
+    def iov_name(i):
+        return f'IOV_{i}'
```

Comment:
This is bad in particular because we cannot run `add_iov` twice on the same `model`.